### PR TITLE
GCW-1895 Twilio upgrade

### PR DESCRIPTION
**Pull request by abulsayyad123**
### Ticket Link: 

https://jira.crossroads.org.hk/browse/GCW-1895

### What does this PR do?

FEATURE: 
1) Upgraded `twilio-ruby` gem to `v5` and its dependency gem `jwt`.
2) Changed token generation logic according to new documentation.
3) updated `TwiML` syntax in `inbound` and `outbound` controllers.
4) Fixed breaking specs according to changes.

### Impacted Areas
* Token generation step.
* Call connect logic.

### Linked PR's:

https://github.com/crossroads/admin.goodcity/pull/358

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -24,11 +24,11 @@ gem 'factory_bot_rails' # used in rake db:seed in production
 gem 'ffaker'
 gem 'execjs'
 # shivani - changed from jwt 0.1.13 to 1.2.0
-gem 'jwt', '~> 1.2.0'
+gem 'jwt', '~> 1.5.0'
 gem 'rack-cors'
 gem 'rack-protection'
 gem 'state_machine'
-gem 'twilio-ruby'
+gem 'twilio-ruby', '~> 5.11.0'
```
Comment (steveyken) [inline Gemfile]: Any reason why we are pinning to a specific version? 
```suggestion
gem 'twilio-ruby', '~> 5'
```

File: Gemfile [hunk 2]
```diff
@@ -24,11 +24,11 @@ gem 'factory_bot_rails' # used in rake db:seed in production
 gem 'ffaker'
 gem 'execjs'
 # shivani - changed from jwt 0.1.13 to 1.2.0
-gem 'jwt', '~> 1.2.0'
+gem 'jwt', '~> 1.5.0'
```
Comment (steveyken) [inline Gemfile]: Not sure we need to pin to a specific version here

File: app/controllers/api/v1/twilio_outbound_controller.rb [hunk 3]
```diff
@@ -44,14 +44,12 @@ class TwilioOutboundController < Api::V1::ApiController
...
+        offer_id, caller_id = params["To"].split("#")
         mobile = Offer.find_by(id: offer_id).created_by.mobile
         TwilioOutboundCallManager.new(to: mobile, offer_id: offer_id, user_id: caller_id).store
-
-        response = Twilio::TwiML::Response.new do |r|
-          r.Dial callerId: voice_number, action: api_v1_twilio_outbound_completed_call_path do |d|
-            d.Number mobile
-          end
+        response = Twilio::TwiML::VoiceResponse.new do |r|
+          r.say(message: "Connecting call to #{user(mobile).full_name}", voice: 'alice')
```
Comment (steveyken) [inline app/controllers/api/v1/twilio_outbound_controller.rb]: nice!

File: config/secrets.yml [hunk 4]
```diff
@@ -3,11 +3,13 @@ base: &BASE
...
     auth_token: <%= ENV['TWILIO_AUTH_TOKEN'] %>
-    phone_number: '<%= ENV['TWILIO_PHONE_NUMBER'] %>'
+    phone_number: <%= ENV['TWILIO_PHONE_NUMBER'] %>
     workflow_sid: <%= ENV['TWILIO_WORKFLOW_SID'] %>
     workspace_sid: <%= ENV['TWILIO_WORKSPACE_SID'] %>
     call_app_sid: <%= ENV['TWILIO_CALL_APP_SID'] %>
-    voice_number: '<%= ENV['TWILIO_VOICE_NUMBER'] %>'
+    voice_number: <%= ENV['TWILIO_VOICE_NUMBER'] %>
+    api_key: <%= ENV['TWILIO_API_KEY'] %>
+    twilio_secret: <%= ENV['TWILIO_SECRET'] %>
```
Comment (steveyken) [inline config/secrets.yml]: Make a note to add these to the .env files

File: Gemfile (hunk 2 above)
Comment (abulsayyad123) [reply Gemfile]: This was the minimum version new `twilio-ruby` needed. Above this version `jwt` was upgrading it's dependent gem.

File: config/secrets.yml (hunk 4 above)
Comment (abulsayyad123) [reply config/secrets.yml]: Sure

**Comment by steveyken** [comment]
Ok that's fine!

**Comment by steveyken** [comment]
All good from my point of view. Keen to test this soon. Let's deploy to master asap

File: spec/controllers/api/v1/twilio_inbound_controller_spec.rb [hunk 5]
```diff
@@ -88,19 +88,22 @@
...
-    end
+    # context "Active Donor" do
+    #   it "will return response to Twilio", :show_in_doc do
+    #     expect(TwilioInboundCallManager).to receive(:new).with(mobile: user.mobile).and_return(double("caller_has_active_offer?" => true, "caller_is_admin?" => false))
+
+    #     post :voice, parameters
+    #     expect(response.status).to eq(200)
+    #     expect(response.body).to eq("<?xml version=\"1.0\" encoding=\"UTF-8\"?><Response><Enqueue workflowSid=\"#{ENV['TWILIO_WORKFLOW_SID']}\" waitUrl=\"/api/v1/twilio_inbound/hold_donor\" waitUrlMethod=\"post\"><TaskAttributes>{\"selected_language\":\"en\",\"user_id\":#{user.id}}</TaskAttributes></Enqueue><Gather numDigits=\"1\" timeout=\"3\" action=\"/api/v1/twilio_inbound/accept_callback\"><Say>Unfortunately none of our staff are able to take your call at the moment.</Say><Say>You can request a call-back without leaving a message by pressing 1.</Say><Say>Otherwise, leave a message after the tone and our staff will get back to you as soon as possible. Thank you.</Say></Gather><Record maxLength=\"60\" playBeep=\"true\" action=\"/api/v1/twilio_inbound/send_voicemail\"/></Response>")
+    #   end
+    # end
```
Comment (swatijadhav) [inline spec/controllers/api/v1/twilio_inbound_controller_spec.rb]: Are we planning to update it later? or can we remove it if not required?

File: app/controllers/api/v1/twilio_inbound_controller.rb [hunk 6]
```diff
@@ -110,9 +110,9 @@ def call_complete
       def call_fallback
         Rollbar.error(Exception, parameters: params,
           error_class: "TwilioError", error_message: "Twilio Voice Call Error")
-        response = Twilio::TwiML::Response.new do |r|
-          r.Say "Unfortunately there is some issue with connecting to Goodcity. Please try again after some time. Thank you."
-          r.Hangup
+        response = Twilio::TwiML::VoiceResponse.new do |r|
+          r.say(message:"Unfortunately there is some issue with connecting to Goodcity. Please try again after some time. Thank you.")
```
Comment (namrataukirde) [inline app/controllers/api/v1/twilio_inbound_controller.rb]: add space after `:` :)

File: spec/controllers/api/v1/twilio_inbound_controller_spec.rb (hunk 5 above)
Comment (abulsayyad123) [reply spec/controllers/api/v1/twilio_inbound_controller_spec.rb]: Need to check about this. If it is not required will remove it in subsequent PR's.

File: app/controllers/concerns/twilio_config.rb [hunk 7]
```diff
@@ -46,10 +46,33 @@ def twilio_client
...
+    api_key = twilio_creds["api_key"]
+    api_secret = twilio_creds["twilio_secret"]
+
+    # Required for Voice
+    outgoing_application_sid = twilio_creds["call_app_sid"]
+    identity = 'user'
+
+    # Create Voice grant for our token
+    grant = Twilio::JWT::AccessToken::VoiceGrant.new
+    grant.outgoing_application_sid = outgoing_application_sid
```
Comment (namrataukirde) [inline app/controllers/concerns/twilio_config.rb]: can't we directly use `twilio_creds["call_app_sid"]` here?

File: app/controllers/api/v1/twilio_inbound_controller.rb (hunk 6 above)
Comment (abulsayyad123) [reply app/controllers/api/v1/twilio_inbound_controller.rb]: Sure will add this change :)

File: app/controllers/concerns/twilio_config.rb (hunk 7 above)
Comment (abulsayyad123) [reply app/controllers/concerns/twilio_config.rb]: Yes.. we can. Just for readability purpose I added it in separate variable similar to where it is been assigned.

**Pull request merged**